bftengine/src/communication/TlsTCPCommunication.cpp

```diff
@@ -245,680 +408,790 @@ class AsyncTlsConnection : public enable_shared_from_this<AsyncTlsConnection> {
+      LOG_DEBUG(_logger,
+                "Verifying server: " << subject << ", " << preverified);
+
+      bool res = check_certificate(
+          cert, "server", string(subject), _expectedDestId);
+      LOG_DEBUG(_logger,
+                "Manual verifying server: " << subject
+                                            << ", authenticated: " << res);
+      return res;
     }
-    X509_NAME_oneline(X509_get_subject_name(cert), subject, 256);
-    LOG_DEBUG("Verifying server: " << subject << ", " << preverified);
-    auto res = check_sertificate(cert, "server", string(subject), _destId);
-    LOG_DEBUG("Manual verifying server: " << subject << ", " << res);
-    return true;
   }
 
   /**
    * certificate pinning
    * check for specific certificate and do not rely on the chain authentication
+   * if verified, it sets explicitly the _destId
    */
-  bool check_sertificate(X509 *receivedCert,
+  bool check_certificate(X509 *receivedCert,
                          string connectionType,
                          string subject,
-                         NodeNum expectedPeerId = -1) {
-    regex r("OU=\\d*", regex_constants::icase);
-    smatch sm;
+                         NodeNum expectedPeerId) {
+    // first, basic sanity test, just to eliminate disk read if the certificate
+    // is unknown.
+    // the certificate must have node id, as we put it in OU field on creation.
+    // since we use pinning we must know who is the remote peer.
+    // peerIdPrefixLength stands for the length of 'OU=' substring
+    size_t peerIdPrefixLength = 3;
+    std::regex r("OU=\\d*", std::regex_constants::icase);
+    std::smatch sm;
     regex_search(subject, sm, r);
-    if (4 > sm.length()) {
-      LOG_ERROR("OU not found or empty: " << subject);
+    if (sm.length() <= peerIdPrefixLength) {
+      LOG_ERROR(_logger, "OU not found or empty: " << subject);
       return false;
     }
 
-    string remPeer = sm.str().substr(0, sm.str().length() - 3);
+    string remPeer =
+        sm.str().substr(
+            peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
     if (0 == remPeer.length()) {
-      LOG_ERROR("OU empty " << subject);
+      LOG_ERROR(_logger, "OU empty " << subject);
       return false;
     }
 
-    int remotePeerId;
+    NodeNum remotePeerId;
     try {
-      remotePeerId = stoi(sm.str());
+      remotePeerId = stoul(remPeer, nullptr);
     } catch (const std::invalid_argument &ia) {
-      LOG_ERROR("cannot convert OU, " << subject);
+      LOG_ERROR(_logger, "cannot convert OU, " << subject << ", " << ia.what());
+      return false;
+    } catch (const std::out_of_range &e) {
+      LOG_ERROR(_logger, "cannot convert OU, " << subject << ", " << e.what());
       return false;
     }
 
-    // if server has been verified, check that we co
-    if (-1 != expectedPeerId) {
+    // if server has been verified, check peers match
+    if (UNKNOWN_NODE_ID != expectedPeerId) {
```

Comment:
The check_certificate() function is used both in client and server. When on server, the expectedPeerId will always be set to UNKNOWN - because we don't know who had connected. In this case there is no need to check if the expected peer id is the actual peer id. This is relevant for the client only. Will add comment.